pkg/server/backend_manager.go

```diff
@@ -81,13 +82,15 @@ var _ Backend = agent.AgentService_ConnectServer(nil)
 type backend struct {
 	// TODO: this is a multi-writer single-reader pattern, it's tricky to
 	// write it using channel. Let's worry about performance later.
-	mu   sync.Mutex // mu protects conn
-	conn agent.AgentService_ConnectServer
+	readLock sync.Mutex
+	recvLock sync.Mutex
+	conn     agent.AgentService_ConnectServer
 }
 
 func (b *backend) Send(p *client.Packet) error {
-	b.mu.Lock()
-	defer b.mu.Unlock()
+	b.readLock.Lock()
+	defer b.readLock.Unlock()
```

Comment:
why is this called `readLock`? I'd expect `sendLock`. It seems weird to guard `Send` with a read lock...

Reply:
I think my brain meant send but fingers typed read. Fixed.